## Dimensional Analysis over Finite Holographic Substrate, [Akhtman, Preprints 2026](https://doi.org/10.20944/preprints202605.0668.v1)

Each cell below verifies one predicate of the paper's predicate ledger (Appendix A; [finitering.space/10-dimensions/](https://finitering.space/10-dimensions/)): it runs the
blocks of [`dimensions.py`](https://finitering.space/src/10-dimensions/) whose checks cite the predicate, prints the check and lists every record that cites it with its verdict. Any cell can be run first, or *Runtime → Run all* (≈ 10 s).

Every cell is self-contained: its first line installs the package from the site (pip reports it already satisfied once it is there), its last runs the predicate.

In [1]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:C2 (p10012) [T] — the deciding family dom.A
# The modular unit-domain group and the grading: $\{U_{r,s}\}$ is a finite abelian group isomorphic to $\Dp$
# under $U_{r,s}U_{r',s'}=U_{r+r',s+s'}$, and $\Ap$ is $\Dp$-graded,
# $\Ap^{(r,s)}\Ap^{(r',s')}\subseteq\Ap^{(r+r',s+s')}$.
from frc_10_dimensions import predicate; predicate("10:C2")

  [PASS] dom.A    EXACT [10:C2, 10:C3, 10:D2, 10:D6, 10:F2, 10:F4, 10:G1, 10:G3, 10:X1, 10:X3] shell datum and domain algebra on F_13: frame datum (g = 2, i = −g^κ, π = 2κ, 4κ = p−1); the group law, invers  --  33 exact micro-checks
  [PASS] dom.B    EXACT [10:E2, 10:E3, 10:X1, 10:X2, 10:X3] the quartet at the unit face: exponent-vector relation lattice over {ℓ_P, t_P, ħ} (both faces of c and ħ diffe  --  30 exact micro-checks
  [PASS] dom.C    EXACT [10:E4] the defining congruences on the lab Carrier Ω = 2 408 561 (S = 602 140): admissibility; 2G ≡ −1 unique; ħ² ≡ −  --  13 exact micro-checks
  [PASS] dom.D    EXACT [10:E4, 10:E5, 10:E9, 10:F3, 10:G1, 10:P2, 10:X2] both Carriers (233 and the lab Carrier): the congruence-and-closure system, ħcG⁻¹ landing in the k_B pair; the  --  51 exact micro-checks
  [PASS] dom.E    EXACT [10:C5] covariance: the naive character ill-defined on the modular projection; (p−1, 0) trivial character yet non-neut  --  18 exact micro-checks
  [PASS] dom.F   

In [2]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:C3 (p10013) [T] — the deciding family dom.A
# Fibrewise addition and the neutral-domain criterion: a sum of homogeneous quantities is homogeneous exactly
# when its nonzero summands share a domain, the fibre components adding separately; a monomial $\prod
# Q_j^{k_j}$ has domain $\sum k_j(r_j,s_j)$ and is a neutral-domain invariant iff that sum is $(0,0)$ in
# $\Dp$.
from frc_10_dimensions import predicate; predicate("10:C3")

— dimensions.py:272 (the check that decides 10:C3; family dom.A)
  272      # 10:C3 (p10013)
  273      chk("squaring leaves fibre: dom(Q) != dom(Q^2) for [L]", L != dpow(L, 2))
  274      H = 5
  275      # 10:G1 (p10037), 10:G3 (p10039)
  276      chk("local recovery bound: p-1 > 2H", P - 1 > 2 * H)
  277      pairs = [(r, s) for r in range(-H, H + 1) for s in range(-H, H + 1)]
  278      images = {(r % P, s % (P - 1)) for (r, s) in pairs}
  279      chk("local recovery: all |r|,|s| <= 5 distinguished", len(images) == len(pairs))
  280      # the energy--momentum relation E^2 = p^2 c^2 + m^2 c^4: its three terms in one fibre, at crossing degree two (the
  281      # half-period sector) -- Example ex:dispersion, predicate 10:G3
  282      E2 = dpow(E_dom, 2); pc2 = dpow(dmul(p_dom, v_dom), 2); mc4 = dpow(dmul(m_dom, dpow(v_dom, 2)), 2)
  283      chk("dispersion: [E]^2 = [pc]^2 = [mc^2]^2, one fibre at sector 2", E2 == pc2 == mc4 and E2[2] == 2)
  284  
  285      # B. the quartet at 

In [3]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:C5 (p10015) [T] — the deciding family dom.E
# Window covariance: on a bounded window $|r|,|s|\le H<2\kap$ in the flag-free sector the labels are covariant
# under every admissible reframing; the bound is sharp (the label $(0,\pi)$ is pushforward-invariant yet
# non-neutral); the naive character is ill-defined on the modular projection and the would-be temporal
# character fails composition ($5^{2}\equiv1$ in $\Z_{12}^{\times}$); on the realized lattice the invariant
# labels are exactly $\{0,2\kap\}$; the $\sigma$-twisted action is equivariant by full sweep on $\p=13$ and
# $229$; the flagged label $(0,0;1)$ moves under $\eps=-1$.
from frc_10_dimensions import predicate; predicate("10:C5")

  [PASS] lift.L1  EXACT [10:D3] operator four-cycle: F = iW on F_p^{p−1}, F² = J, F⁴ = I, F² ≠ I (p = 13, 173)  --  2 exact micro-checks
  [PASS] lift.L2  EXACT [10:D3] chart shadow of order two: F exchanges the two dual charts, J exchanges none; the cardinal skeleton acts on ch  --  2 exact micro-checks
  [PASS] lift.L3  EXACT [10:D3] record map s ↦ I_q^s an isomorphism Z_4 → the flag subgroup {0, κ, 2κ, 3κ}; the quotient by {0, 2κ} returns s   --  2 exact micro-checks
  [PASS] lift.L4  EXACT [10:D3] Carrier face: ħ² = −1 (mod Ω), ħ⁴ = 1, ħ² ≠ 1 — the crossing quantum has order four, never two (Ω = 233, 2 408  --  2 exact micro-checks
  [PASS] lift.L5  EXACT [10:C5, 10:D3] invariance on the realized lattice: {u : εu = u for every admissible ε} = {0, 2κ}; witnesses (κ,1) ↦ 2κ and (−  --  2 exact micro-checks
— dimensions.py:556 (the check that decides 10:C5; family dom.E)
  556      # 10:C5 (p10015)
  557      # window covariance (Theorem 5, repaired form)
  558      H = 5
  559      c

In [4]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:D1 (p10017) [T] — the deciding family dom.F
# Chart duality: $[f]=\unitT^{-1}$; $\delta_S:D\mapsto D^{-1}$ is an involution of the domain group whose
# orbit on the generators is the four-domain structure (space, momentum, time, energy); $\delta_C$ the second
# involution, carrying $\unitL\mapsto[p]$, $\unitT\mapsto\unitE$.
from frc_10_dimensions import predicate; predicate("10:D1")

— dimensions.py:625 (the check that decides 10:D1; family dom.F)
  625      # 10:D1 (p10017)
  626      # delta_S / delta_C: involutions; delta_C carries primal domains to dual horizon domains
  627      def dS_(a):
  628          return dinv(a)
  629      def dC_(a):
  630          return dmul(FLAG, dinv(a))
  631      chk("r03: delta_S involution", all(dS_(dS_(a)) == a for a in [L, T, FLAG, (3, 5, 2)]))
  632      chk("r03: delta_C involution", all(dC_(dC_(a)) == a for a in [L, T, FLAG, (3, 5, 2)]))
  633      chk("r03: delta_C[L] = [p] domain", dC_(L) == p_dom)
  634      chk("r03: delta_C[T] = [E] domain", dC_(T) == E_dom)
  635      chk("r03: delta_S flag-free", dS_(L)[2] == 0 and dS_(T)[2] == 0)
  636  
  637  
  638      # G. round-04 layer: realized action, j-restriction, ladder, k_B linkage typing
  [PASS] dom.F    (deciding)       15 exact micro-checks
10:D1: VERIFIED — 1 family record(s)


In [5]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:D2 (p10018) [T] — the deciding family dom.A
# The internal flag: $\Dp$ contains exactly one subgroup of order four, $\langle\unitT^{\kap}\rangle$,
# entirely in the time-exponent factor; $\Iq:=\unitT^{\kap}$, $\Iq^{4}=1$, $\Iq^{2}=\unitT^{\pi}$; no flag of
# space ($\Z_\p$ has no element of order four); the two periods interfere, $(\unitL\unitT^{\kap})^{\p}=\Iq$,
# on three shells.
from frc_10_dimensions import predicate; predicate("10:D2")

— dimensions.py:214 (the check that decides 10:D2; family dom.A)
  214      # 10:D2 (p10018)
  215      # fork B: the flag is internal, Iq = [T]^kappa; realized label of (r,s;j) is
  216      # (r mod p, (s + j*kappa) mod (p-1)) with sector j mod 4
  217      def realize(a):
  218          return (a[0] % P, (a[1] + a[2] * KAPPA) % (P - 1))
  219      chk("realized flag label = (0, kappa)", realize(FLAG) == (0, KAPPA))
  220      o4 = [(r, s) for r in range(P) for s in range(P - 1)
  221            if (4 * r) % P == 0 and (4 * s) % (P - 1) == 0
  222            and not ((2 * r) % P == 0 and (2 * s) % (P - 1) == 0)]
  223      # 10:X3 (p10050)
  224      chk("unique order-four subgroup: exactly two generators", sorted(o4) == [(0, KAPPA), (0, 3 * KAPPA)])
  225      chk("no flag of space: all order-four elements have r = 0", all(r == 0 for r, _ in o4))
  226      chk("flag horizon-inaccessible on window H<kappa",
  227            all((0, KAPPA) != (r % P, s % (P - 1))
  [PASS] dom.A    (d

In [6]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:D3 (p10019) [T] — the deciding family lift.L1
# The lift: the flag records the quarter the chart duality forgets --- on charts the cardinal skeleton acts as
# $s\bmod2$ ($F$ exchanges the conjugate charts, $J$ fixes them), on labels $s\mapsto\Iq^{s}$ is an
# isomorphism $\Z_4\to\langle\Iq\rangle$ whose quotient by $\{0,2\kap\}$ returns the chart action; on the
# Carrier $\hbar^{2}=-1$, $\hbar^{4}=1$, $\hbar^{2}\ne1$, the crossing quantum of order four, never two.
# Shells $(13,2)$, $(173,3)$; Carriers $233$, $2\,408\,561$.
from frc_10_dimensions import predicate; predicate("10:D3")

— dimensions.py:813 (the check that decides 10:D3; family lift.L1)
  813      # 10:D3 (p10019)
  814      # shell Fourier matrix W_{jk} = g^{jk} on F_p^{n}; F = i*W
  815      W = [[pow(g, (j * l) % n, p) for l in range(n)] for j in range(n)]
  816      F = [[(i_res * W[j][l]) % p for l in range(n)] for j in range(n)]
  817  
  818      F2 = modmat_mul(F, F, p)
  819      F4 = modmat_mul(F2, F2, p)
  820      I = [[1 if a == b else 0 for b in range(n)] for a in range(n)]
  821      # parity J: index inversion j -> -j mod n
  822      J = [[1 if b == (-a) % n else 0 for b in range(n)] for a in range(n)]
  823  
  824      l1 = (F2 == J) and (F4 == I) and (F2 != I)
  825  
  826      # L2: chart shadow. Charts = the primal/dual pair; F steps the cardinal
  [PASS] lift.L1  (deciding)       2 exact micro-checks
  [PASS] lift.L2  (corroborating)  2 exact micro-checks
  [PASS] lift.L3  (corroborating)  2 exact micro-checks
  [PASS] lift.L4  (corroborating)  2 exact micro-checks
  [PASS] lift

In [7]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:D6 (p10022) [T] — the deciding family dom.A
# The flagged readings over the counts: $\unitE=[h][f]=\Iq\unitT^{-1}$, $[p]=\unithbar[k]=\Iq\unitL^{-1}$, one
# quantity per pair, the flagged reading one capacity step above its count; $[S]=\unitE\unitT=\Iq$ and every
# phase count neutral, $\unitE\unitT\unithbar^{-1}=1$; the absolute unit is horizon-inaccessible (no monomial
# of a window $H\ll\kap$ expresses the flag) and becomes visible at $\kap\le H<2\kap$.
from frc_10_dimensions import predicate; predicate("10:D6")

— dimensions.py:246 (the check that decides 10:D6; family dom.A)
  246      # 10:D6 (p10022)
  247      E_dom = dmul(FLAG, dinv(T))                       # [E] = Iq T^-1
  248      p_dom = dmul(FLAG, dinv(L))                       # [p] = Iq L^-1
  249      v_dom = dmul(L, dinv(T))                          # [v] = L T^-1
  250      m_dom = dmul(E_dom, dinv(dpow(v_dom, 2)))         # [m] = [E][v]^-2
  251      chk("[m] = Iq L^-2 T", m_dom == ((-2) % P, 1, 1))
  252      F_dom = dmul(m_dom, dmul(L, dpow(dinv(T), 2)))    # [F] = [m][a]
  253      chk("[F] = Iq L^-1 T^-1", F_dom == ((-1) % P, (-1) % (P - 1), 1))
  254      S_dom = dmul(E_dom, T)
  255      chk("[S] = Iq (action carries the flag)", S_dom == FLAG)
  256      chk("phase count neutral: [E][T][hbar]^-1 = 1", dmul(S_dom, dinv(FLAG)) == ONE)
  257      G_dom = dmul(F_dom, dmul(dpow(L, 2), dinv(dpow(m_dom, 2))))
  258      chk("[G] = Iq^-1 L^5 T^-3", G_dom == (5, (-3) % (P - 1), 3))
  259      chk("Planck area flag-free: [G][hbar]

In [8]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:E2 (p10025) [T] — the deciding family dom.B
# One crossing, one action, and pairing closure: $\lP\pP=\tPl\EP=\hbar$ (the two faces of $c$ and $\hbar$
# differ by the single relation $\ell p=tE$); the cross-domain relations reduce to the two generators $c$,
# $\hbar$, the mixed products their monomials, $|k_B|=\hbar/c$ derived; the relation lattice over
# $\{\lP,\tPl,\hbar\}$ has rank two and $k_B$, $\lP\EP$ add nothing; $4-1=3$ free scales in bijection with
# $\{c,\hbar,G\}$, the classical arity derived.
from frc_10_dimensions import predicate; predicate("10:E2")

— dimensions.py:325 (the check that decides 10:E2; family dom.B)
  325      # 10:E2 (p10025)
  326      # pairing lattice mod the relation has rank TWO: mix = hbar - c, lE = hbar + c
  327      chk("mix = hbar' - c' (exponent identity: k_B derived)",
  328            mix_vec == sub(hbar_vec2, c_vec2))
  329      chk("pairing rank 2 mod relation: rank{c,hbar,rel} = 3",
  330            rank_int([c_vec, hbar_vec, rel]) == 3)
  331      chk("mix adds nothing: rank{c,hbar,mix,rel} = 3",
  332            rank_int([c_vec, hbar_vec, mix_vec, rel]) == 3)
  333      chk("lE adds nothing: rank{c,hbar,mix,lE,rel} = 3",
  334            rank_int([c_vec, hbar_vec, mix_vec, lE_vec, rel]) == 3)
  335  
  336      # 10:E3 (p10026)
  337      # numeric instantiation with the identity enforced: choose l,t,p free, E = l*p/t
  338      l, t, p = Fr(3, 7), Fr(2, 5), Fr(11, 4)
  [PASS] dom.B    (deciding)       30 exact micro-checks
10:E2: VERIFIED — 1 family record(s)


In [9]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:E3 (p10026) [T] — the deciding family dom.B
# The cancellation identity $|k_B|c=\hbar$ at the unit face, $(\pP\tPl)(\lP/\tPl)=\lP\pP$; mass derived,
# $m_P=\pP/c=\EP/c^{2}$; the temperature horizon $\Theta_P=\EP/|k_B|$; the flag positions $(0,0,1,1)$ and the
# closure of every quartet relation on the flag component; no $\hbar$ of space and time.
from frc_10_dimensions import predicate; predicate("10:E3")

— dimensions.py:336 (the check that decides 10:E3; family dom.B)
  336      # 10:E3 (p10026)
  337      # numeric instantiation with the identity enforced: choose l,t,p free, E = l*p/t
  338      l, t, p = Fr(3, 7), Fr(2, 5), Fr(11, 4)
  339      E = l * p / t
  340      c_u, hbar_u = l / t, l * p
  341      chk("unit face: c = l/t = E/p", c_u == E / p)
  342      chk("unit face: hbar = l p = t E", hbar_u == t * E)
  343      chk("cancellation: (p t)(l/t) = l p  [k_B c = -hbar]", (p * t) * (l / t) == hbar_u)
  344      chk("mixed partner: l E = hbar c", l * E == hbar_u * c_u)
  345      m_P = p / c_u
  346      chk("m_P = p/c = E/c^2", m_P == E / c_u ** 2)
  347      G_num = hbar_u * c_u / m_P ** 2
  348      chk("G = hbar c / m_P^2 = l^2 c^3 / hbar", G_num == l ** 2 * c_u ** 3 / hbar_u)
  349      # bijection {c, hbar, G} <-> quartet: recover the free scale from G
  [PASS] dom.B    (deciding)       30 exact micro-checks
10:E3: VERIFIED — 1 family record(s)


In [10]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:E4 (p10027) [T] — the deciding family dom.C
# The defining congruences, pair form: on the Carrier chart $2G+1\equiv0$, $2c^{2}\equiv1$,
# $\hbar^{2}\equiv-1$, $k_B^{2}\equiv-2\pmod\Om$ determine the constants uniquely at pair level --- $G=2\dS$
# exact, the quadratics each with exactly two roots $\{x,-x\}$ (admissibility guarantees the residues:
# $\Om\equiv1\bmod4$, $\dS$ even for $2$, hence $-2$) --- and the linkage $\{\pm k_B\}\{\pm c\}=\{\pm\hbar\}$
# is derived at pair level; the root pair of $-1$ is $\{\hbar,h\}$ with $h=2\pi\hbar\equiv-\hbar$. Exhaustive
# on both Carriers.
from frc_10_dimensions import predicate; predicate("10:E4")

— dimensions.py:403 (the check that decides 10:E4; family dom.C)
  403      # 10:E4 (p10027)
  404      # congruence 1: 2G + 1 = 0 -- linear, unique
  405      G_pin = [x for x in (pow(2, -1, Om) * (Om - 1) % Om,) ]
  406      G_val = (Om - 1) // 2
  407      chk("congruence 2G+1=0 unique: G = 2S = 1,204,280", (2 * G_val + 1) % Om == 0 and G_val == 2 * S == 1_204_280)
  408  
  409      # quadratic defining congruences: exhaustive-root verification via the known values
  410      hbar_val = 18_688
  411      kB_val = 1_880_160
  412      chk("congruence hbar^2 = -1", (hbar_val * hbar_val) % Om == Om - 1)
  413      chk("congruence k_B^2 = -2", (kB_val * kB_val) % Om == Om - 2)
  414      c2 = (2 * S + 1) % Om
  415      chk("congruence 2c^2 = 1: c^2 = 2S+1", (2 * c2) % Om == 1)
  416      # each quadratic congruence has exactly two roots: x and Om - x
  [PASS] dom.C    (deciding)       13 exact micro-checks
  [PASS] dom.D    (corroborating)  51 exact micro-checks
  [PASS] dom.H    (cor

In [11]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:E5 (p10028) [T] — the deciding family dom.H
# Pair consequences and representative inertness: $G=-c^{2}$, $G^{2}\equiv4^{-1}$, $\hbar^{4}\equiv1$,
# $(k_Bc)^{2}\equiv-1$, and $\hbar cG^{-1}$ lands in $\{\pm k_B\}$; of the eight sign assignments exactly the
# four with $\sigma_\hbar=\sigma_c\sigma_k$ are admissible, a $(\Z/2)^{2}$, every identity holding on each and
# the $\hbar$-flip relabelling within $\{\hbar,h\}$; what a bounded observer registers is exactly the
# pair-inert content.
from frc_10_dimensions import predicate; predicate("10:E5")

— dimensions.py:754 (the check that decides 10:E5; family dom.H)
  754          # 10:E5 (p10028)
  755      # representative inertness: exactly the assignments with s_h = s_c*s_k are
  756          # admissible (a (Z/2)^2 group), and every checked identity holds on each
  757          admissible_count = 0
  758          for s_c in (1, -1):
  759              for s_h in (1, -1):
  760                  for s_k in (1, -1):
  761                      hb2, c2_, kB2 = (s_h * R["hbar"]) % Om_, (s_c * R["c"]) % Om_, (s_k * R["kB"]) % Om_
  762                      quad = ((2 * c2_ * c2_) % Om_ == 1 and (hb2 * hb2) % Om_ == Om_ - 1
  763                              and (kB2 * kB2) % Om_ == Om_ - 2)
  764                      link = (kB2 * c2_) % Om_ == (Om_ - hb2) % Om_
  765                      if quad and link:
  766                          admissible_count += 1
  767                          h2 = (Om_ - hb2) % Om_
  [PASS] dom.D    (corroborating)  51 exact micro-checks
  [PASS] dom.H    

In [12]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:E9 (p10032) [T] — the deciding family dom.D
# The minimal admissible pair: under the programme's admissibility predicate ($\p=4\kap+1$ prime, $\kap>1$;
# $\Om=4\dS+1$ prime, $\dS$ even, $\dS\equiv1\bmod3$; $\p^{2}<\Om$) the smallest instance is
# $(\p,\Om)=(13,233)$, by exhaustive scan, with the counterfactuals: dropping the mod-$3$ clause admits
# $(13,193)$, dropping $\kap>1$ admits $(5,41)$, $\kap=2$ gives the composite $9$.
from frc_10_dimensions import predicate; predicate("10:E9")

— dimensions.py:487 (the check that decides 10:E9; family dom.D)
  487      # 10:E9 (p10032)
  488      # minimality scan under the complete predicate (the paper's Definitions 1 and 2, Remark 1)
  489      def admissible(p_, Om__):
  490          kap_ = (p_ - 1) // 4
  491          S__ = (Om__ - 1) // 4
  492          return (kap_ > 1 and is_prime(p_) and p_ % 4 == 1
  493                  and Om__ % 4 == 1 and is_prime(Om__)
  494                  and S__ % 2 == 0 and S__ % 3 == 1
  495                  and p_ * p_ < Om__)
  496      adm = [(p_, Om__) for p_ in range(5, 40, 4) for Om__ in range(p_ * p_ + 1, 234)
  497             if admissible(p_, Om__)]
  498      chk("minimality: (13,233) admissible", (13, 233) in adm)
  499      chk("minimality: no admissible pair below 233", min(o for _, o in adm) == 233)
  500      chk("counterfactual: dropping mod-3 admits (13,193)",
  [PASS] dom.D    (deciding)       51 exact micro-checks
10:E9: VERIFIED — 1 family record(s)


In [13]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:F2 (p10034) [T] — the deciding family dom.A
# The mechanical and gravitational domains: $[m]=\Iq\unitL^{-2}\unitT$; $[a]=\unitL\unitT^{-2}$,
# $[F]=\Iq\unitL^{-1}\unitT^{-1}$, $[p]=\Iq\unitL^{-1}$, $[S]=\Iq$, $[P]=\Iq\unitT^{-2}$,
# $[\text{pressure}]=\Iq\unitL^{-3}\unitT^{-1}$; $[G]=\Iq^{-1}\unitL^{5}\unitT^{-3}$; the geometric
# conversions $[G\hbar/c^{3}]=\unitL^{2}$, $[Gm/c^{2}]=\unitL$, $[Gm/c^{3}]=\unitT$, $[G\rho_m]=\unitT^{-2}$;
# mass derived, not primitive.
from frc_10_dimensions import predicate; predicate("10:F2")

— dimensions.py:237 (the check that decides 10:F2; family dom.A)
  237      # 10:F2 (p10034)
  238      # derived domains
  239      def dpow(a, k):
  240          out = ONE
  241          step = a if k >= 0 else dinv(a)
  242          for _ in range(abs(k)):
  243              out = dmul(out, step)
  244          return out
  245  
  246      # 10:D6 (p10022)
  247      E_dom = dmul(FLAG, dinv(T))                       # [E] = Iq T^-1
  248      p_dom = dmul(FLAG, dinv(L))                       # [p] = Iq L^-1
  249      v_dom = dmul(L, dinv(T))                          # [v] = L T^-1
  250      m_dom = dmul(E_dom, dinv(dpow(v_dom, 2)))         # [m] = [E][v]^-2
  [PASS] dom.A    (deciding)       33 exact micro-checks
10:F2: VERIFIED — 1 family record(s)


In [14]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:F3 (p10035) [T] — the deciding family dom.D
# Temperature carries the acceleration domain: $[\Theta]=\unitE[k_B]^{-1}=\unitL\unitT^{-2}=[a]$, flag-free;
# the Unruh combination $\hbar a/c\,k_B$ closes flag-free in the algebra; neither mass nor temperature is
# primitive, the classical $M$-$L$-$T$-plus-thermal system the torsion-free shadow of two generators plus the
# flag.
from frc_10_dimensions import predicate; predicate("10:F3")

— dimensions.py:513 (the check that decides 10:F3; family dom.D)
  513      # 10:F3 (p10035), 10:X2 (p10049)
  514      # temperature and Unruh closures in (r, s, j) bookkeeping
  515      kB_dom = (-1, 1, 1)                    # Iq L^-1 T
  516      E_dom3 = (0, -1, 1)                    # Iq T^-1
  517      acc = (1, -2, 0)
  518      Theta = tuple(a - b for a, b in zip(E_dom3, kB_dom))
  519      chk("[Theta] = [E][k_B]^-1 = L T^-2, flag-free", Theta == acc)
  520      c_dom3 = (1, -1, 0)
  521      hb_dom3 = (0, 0, 1)
  522      unruh = tuple(h + a - c - k for h, a, c, k in zip(hb_dom3, acc, c_dom3, kB_dom))
  523      chk("Unruh closure: [hbar a / (c k_B)] = [Theta]", unruh == acc)
  524  
  525      # crossing-degree recovery (Cor recovery-crossings)
  526      def embed(u, a, b):
  [PASS] dom.D    (deciding)       51 exact micro-checks
10:F3: VERIFIED — 1 family record(s)


In [15]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:F4 (p10036) [T] — the deciding family dom.A
# Count-valued comparisons are flag-free: every ratio of two quantities of equal crossing degree ($Gm/r^{3}$,
# the phase exponent $[E][T]/[\hbar]$) is a neutral or flag-free label, the flag entering unitful measures
# only; $F/a=[m]$, crossing degrees $1$ and $0$, is the flagged boundary case.
from frc_10_dimensions import predicate; predicate("10:F4")

— dimensions.py:265 (the check that decides 10:F4; family dom.A)
  265      # 10:F4 (p10036)
  266      chk("[Gm/r^3] = T^-2 flag-free",
  267            dmul(G_dom, dmul(m_dom, dpow(dinv(L), 3))) == (0, (-2) % (P - 1), 0))
  268      chk("Compton flag-free: [m][c][hbar]^-1 = L^-1",
  269            dmul(m_dom, dmul(v_dom, dinv(FLAG))) == ((-1) % P, 0, 0))
  270  
  271      # fibrewise addition / neutral criterion / local recovery
  272      # 10:C3 (p10013)
  273      chk("squaring leaves fibre: dom(Q) != dom(Q^2) for [L]", L != dpow(L, 2))
  274      H = 5
  275      # 10:G1 (p10037), 10:G3 (p10039)
  276      chk("local recovery bound: p-1 > 2H", P - 1 > 2 * H)
  277      pairs = [(r, s) for r in range(-H, H + 1) for s in range(-H, H + 1)]
  278      images = {(r % P, s % (P - 1)) for (r, s) in pairs}
  [PASS] dom.A    (deciding)       33 exact micro-checks
10:F4: VERIFIED — 1 family record(s)


In [16]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:F5 (p10047) [T|D] — the deciding family dom.G
# The electromagnetic domain: with the Coulomb constant neutral (declared),
# $[q]^{2}=\unitE\unitL=\Iq\unitL\unitT^{-1}$; charge is an integer degree on the lift realised as
# $\unitL^{2\kap+1}\unitT^{(\kap-1)/2}$, existing iff $\kap$ is odd, unique up to $\Iq^{2}$, never
# window-covariant ($|r|=2\kap$); no further generator within C1's two charts; on the even capacities no label
# squares to $[q]^{2}$.
from frc_10_dimensions import predicate; predicate("10:F5")

— dimensions.py:709 (the check that decides 10:F5; family dom.G)
  709      # 10:F5 (p10047), 10:X4 (p10051)
  710      chk("charge: [q]^2 = [E][L] = I_q [L][T]^-1 = (1, 2) on F_13, its roots exactly (7,1) and (7,7), differing by the half-period (0, 6)",
  711            ((0 + 1) % 13, (3 - 1 + 0) % 12) == (1, 2) and charge_roots(3) == [(7, 1), (7, 7)] and ((7 - 7) % 13, (7 - 1) % 12) == (0, 6))
  712      chk("charge: the roots exist iff kappa is odd, then exactly (2 kappa+1, (kappa-1)/2) and its half-period partner (every prime p = 4 kappa+1, kappa <= 30)",
  713            all((charge_roots(k) == sorted([(2 * k + 1, (k - 1) // 2), (2 * k + 1, (k - 1) // 2 + 2 * k)])) if k % 2 else charge_roots(k) == []
  714                for k in range(1, 31) if all((4 * k + 1) % d for d in range(2, 4 * k + 1))))
  715      chk("charge: the space exponent 2^-1 = 2 kappa+1 = -2 kappa is the meridian half-turn, |r| = 2 kappa, on p = 13, 29, 173, 229",
  716            all((2 * (2 * k + 1)) % (4 * k 

In [17]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:G1 (p10037) [T] — the deciding family dom.A
# Local recovery: for an exponent horizon $H$ with $4\kap>2H$ the modular labels distinguish every
# conventional pair $|r|,|s|\le H$; the flagged sector is recovered through the crossing degree,
# $M^{u}L^{a}T^{b}\mapsto(a-2u,\,b+u;\,u)$ injective and windowed-faithful.
from frc_10_dimensions import predicate; predicate("10:G1")

— dimensions.py:275 (the check that decides 10:G1; family dom.A)
  275      # 10:G1 (p10037), 10:G3 (p10039)
  276      chk("local recovery bound: p-1 > 2H", P - 1 > 2 * H)
  277      pairs = [(r, s) for r in range(-H, H + 1) for s in range(-H, H + 1)]
  278      images = {(r % P, s % (P - 1)) for (r, s) in pairs}
  279      chk("local recovery: all |r|,|s| <= 5 distinguished", len(images) == len(pairs))
  280      # the energy--momentum relation E^2 = p^2 c^2 + m^2 c^4: its three terms in one fibre, at crossing degree two (the
  281      # half-period sector) -- Example ex:dispersion, predicate 10:G3
  282      E2 = dpow(E_dom, 2); pc2 = dpow(dmul(p_dom, v_dom), 2); mc4 = dpow(dmul(m_dom, dpow(v_dom, 2)), 2)
  283      chk("dispersion: [E]^2 = [pc]^2 = [mc^2]^2, one fibre at sector 2", E2 == pc2 == mc4 and E2[2] == 2)
  284  
  285      # B. the quartet at the unit face (exact rationals over free scales)
  286      family('dom', 'B')
  287      # Represent horizons as exponent vectors

In [18]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:G2 (p10038) [T] — the deciding family dom.G
# The window ladder: coherence $2\sqrt\kap$ below recovery $\kap/2$ below flag inaccessibility $\kap$ below
# covariance $2\kap$, nested strictly for every $\kap\ge17$ and failing for the toy $\kap=3$, all orderings by
# integer squares; the coherence identity $(2\sqrt\kap)^{2}=\p-1$ and the totality closure
# $(2\sqrt\dS)^{2}=\Om-1$.
from frc_10_dimensions import predicate; predicate("10:G2")

— dimensions.py:650 (the check that decides 10:G2; family dom.G)
  650      # 10:G2 (p10038)
  651      # window ladder: nested for kappa >= 17; toy fails.  All orderings exact:
  652      # 2 sqrt(k) < k/2  <=>  (4 sqrt(k))^2 < k^2  <=>  16 k < k^2  <=>  k > 16.
  653      def ladder(k):
  654          return 16 * k < k * k and 0 < k
  655      chk("r04: ladder nested for kappa=17,387,602140 (integer-square ordering)",
  656            all(ladder(k) for k in (17, 387, 602140)))
  657      chk("r04: toy kappa=3 below nesting threshold", not ladder(3))
  658      chk("r04: coherence window identity (2 sqrt kappa)^2 = 4 kappa = p-1 exactly",
  659            all((2 * 2 * k == 4 * k) and (4 * k == (4 * k + 1) - 1) for k in (3, 387, 602140)))
  660      chk("r04: totality closure exact: (2 sqrt S)^2 = Om - 1",
  661            4 * 602140 == 2408561 - 1)
  662  
  663      # 10:G5 (p10052), 10:P1 (p10053)
  [PASS] dom.G    (deciding)       29 exact micro-checks
10:G2: VERIFIED — 1 family re

In [19]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:G3 (p10039) [T] — the deciding family dom.A
# The worked examples: $\kap=3$, $H=5$, $12>10$ so every pair in $[-5,5]^{2}$ is distinguished; kinetic energy
# $[m][v]^{2}=[E]$; $Q+Q^{2}$ inhomogeneous; the phase exponent neutral; flag arithmetic ($G\hbar/c^{3}$
# flag-free, $\Iq^{2}=\unitT^{\pi}$); the Schwarzschild length $[Gm/c^{2}]=\unitL$; the gravitational
# frequency $[Gm/r^{3}]=\unitT^{-2}$; the energy--momentum relation $E^{2}=p^{2}c^{2}+m^{2}c^{4}$ homogeneous
# at crossing degree two, its massless case $E=pc$ at crossing degree one.
from frc_10_dimensions import predicate; predicate("10:G3")

— dimensions.py:275 (the check that decides 10:G3; family dom.A)
  275      # 10:G1 (p10037), 10:G3 (p10039)
  276      chk("local recovery bound: p-1 > 2H", P - 1 > 2 * H)
  277      pairs = [(r, s) for r in range(-H, H + 1) for s in range(-H, H + 1)]
  278      images = {(r % P, s % (P - 1)) for (r, s) in pairs}
  279      chk("local recovery: all |r|,|s| <= 5 distinguished", len(images) == len(pairs))
  280      # the energy--momentum relation E^2 = p^2 c^2 + m^2 c^4: its three terms in one fibre, at crossing degree two (the
  281      # half-period sector) -- Example ex:dispersion, predicate 10:G3
  282      E2 = dpow(E_dom, 2); pc2 = dpow(dmul(p_dom, v_dom), 2); mc4 = dpow(dmul(m_dom, dpow(v_dom, 2)), 2)
  283      chk("dispersion: [E]^2 = [pc]^2 = [mc^2]^2, one fibre at sector 2", E2 == pc2 == mc4 and E2[2] == 2)
  284  
  285      # B. the quartet at the unit face (exact rationals over free scales)
  286      family('dom', 'B')
  287      # Represent horizons as exponent vectors

In [20]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:G4 (p10046) [T] — the deciding family dom.G
# Buckingham's count: on the integer lift the neutral monomials of $N$ quantities are the kernel of their
# $N\times3$ label matrix, $N-\operatorname{rank}$ independent dimensionless products with rank $\le3$, the
# classical count for quantities whose classical dimensions lie in $M$-$L$-$T$ and the derived-$k_B$ count for
# thermal ones; on the window $2H<\kap$, $|j|\le1$ lifted and realized neutrality coincide; the pendulum
# $(T,\ell,g,m)$: rank $3$, one product.
from frc_10_dimensions import predicate; predicate("10:G4")

— dimensions.py:700 (the check that decides 10:G4; family dom.G)
  700      # 10:G4 (p10046)
  701      chk("Buckingham: pendulum (T, l, g, m) rank 3, one product T^2 g/l with label (0,0;0)", rank_int(pend) == 3 and lab_ == (0, 0, 0))
  702      # The electromagnetic domain (Remark rem:electromagnetic, predicate 10:F5): with the Coulomb constant neutral, [q]^2 = [E][L] =
  703      # I_q [L][T]^-1, on the shell (1, kappa-1), on the lift (1,-1;1). Its square roots in D_p = Z_p x Z_{4 kappa}: the space half
  704      # 2^-1 = 2 kappa + 1 = -2 kappa (the meridian half-turn); the phase equation 2s = kappa-1 (mod 4 kappa) is solvable iff kappa
  705      # is odd, then two roots differing by the half-period (0, 2 kappa) = I_q^2; on the lift no root at all (odd coordinates).
  706      def charge_roots(k):
  707          p_, n_ = 4 * k + 1, 4 * k
  708          return sorted((r, s) for r in range(p_) for s in range(n_) if ((2 * r) % p_, (2 * s) % n_) == (1, (k - 1) % n_))
  709      # 10:F5

In [21]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:G5 (p10052) [T] — the deciding family dom.G
# The resolution floor: a registration resolved to one part in $N$ is a coherent count of $N$, bounded at
# $\sqrt\p$ (A5), so a single coherent registration on a shell of cardinality $\p$ resolves no finer than
# $\p^{-1/2}$; under $\p^{2}<\Om$ the floor is at least $\Om^{-1/4}$ on every admissible shell, of the order
# of $10^{-31}$ ($102$ bits) at $\dS=2.6\times10^{122}$ (Rem.~\ref{rem:holographic}), thirteen orders below
# the $10^{-18}$ optical-clock comparisons and twenty-six below the uncertainty of $G$.
from frc_10_dimensions import predicate; predicate("10:G5")

— dimensions.py:663 (the check that decides 10:G5; family dom.G)
  663      # 10:G5 (p10052), 10:P1 (p10053)
  664      # the resolution floor (Remark rem:measured; the prediction P1): a single coherent registration on a shell of cardinality p resolves no finer
  665      # than p^(-1/2) (coherent counts bounded at sqrt p, the import A5); under p^2 < Om the floor is at least Om^(-1/4) on every
  666      # admissible shell. On the chart value S = 2.6e122 (A2), Om = 4S+1: Om^(-1/4) is of the order of 1e-31
  667      # (1e-31 <= Om^(-1/4) < 1e-30  <=>  1e120 < Om <= 1e124); the ratio 1e-18 / Om^(-1/4) rounds to thirteen orders
  668      # (1e12.5 <= Om^(1/4) 1e-18 < 1e13.5  <=>  1e122 <= Om < 1e126); the ratio 2.2e-5 / Om^(-1/4) rounds to twenty-six
  669      # (1e25.5 <= 2.2e-5 Om^(1/4) < 1e26.5  <=>  1e126 <= 22^4 Om < 1e130). Every comparison exact in integers.
  670      S_chart = 26 * 10**121; Om_chart = 4 * S_chart + 1
  671      chk("r04: floor Om^(-1/4) of order 1e-31 on the c

In [22]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:P1 (p10053) [T|R] — the deciding family dom.G
# The resolution floor, staked: no single coherent registration resolves finer than $\Om^{-1/4}$, of the order
# of $10^{-31}$, and no coherent count exceeds $\Om^{1/4}\approx6\times10^{30}$ (G5 over A2, A5, B1; the
# holographic reading of Rem.~\ref{rem:holographic}). Falsifier: one coherent registration finer than that.
# Confrontation: the finest interspecies frequency ratio, $3.2\times10^{-18}$ \citep{Aeppli2025}, and the
# finest differential comparison, $7.6\times10^{-21}$ \citep{Bothwell2022} --- $67$ bits against $102$,
# averaged determinations; the mass-independent clock bound $(t_P^{2}t)^{1/3}$ permits finer beyond $730$~s.
from frc_10_dimensions import predicate; predicate("10:P1")

— dimensions.py:663 (the check that decides 10:P1; family dom.G)
  663      # 10:G5 (p10052), 10:P1 (p10053)
  664      # the resolution floor (Remark rem:measured; the prediction P1): a single coherent registration on a shell of cardinality p resolves no finer
  665      # than p^(-1/2) (coherent counts bounded at sqrt p, the import A5); under p^2 < Om the floor is at least Om^(-1/4) on every
  666      # admissible shell. On the chart value S = 2.6e122 (A2), Om = 4S+1: Om^(-1/4) is of the order of 1e-31
  667      # (1e-31 <= Om^(-1/4) < 1e-30  <=>  1e120 < Om <= 1e124); the ratio 1e-18 / Om^(-1/4) rounds to thirteen orders
  668      # (1e12.5 <= Om^(1/4) 1e-18 < 1e13.5  <=>  1e122 <= Om < 1e126); the ratio 2.2e-5 / Om^(-1/4) rounds to twenty-six
  669      # (1e25.5 <= 2.2e-5 Om^(1/4) < 1e26.5  <=>  1e126 <= 22^4 Om < 1e130). Every comparison exact in integers.
  670      S_chart = 26 * 10**121; Om_chart = 4 * S_chart + 1
  671      chk("r04: floor Om^(-1/4) of order 1e-31 on the c

In [23]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:P2 (p10054) [T|R] — the deciding family dom.D
# Constancy, staked: every dimensionless ratio of the constants is fixed in time and position, zero drift,
# zero dipole --- the totality timeless (B1), every constant a residue of it in the framework's realisation
# (A1; the quartet's, E1, E6; the charge's, Rem.~\ref{rem:electromagnetic}), the quartet's fixed by the
# congruences given $\dS$ (E4), selected once (E7). Falsifier: a detected variation (E1's). Confrontation:
# $\dot\alpha/\alpha=1.8(2.5)\times10^{-19}\,\mathrm{yr}^{-1}$ \citep{Filzinger2023}, consistent with zero.
from frc_10_dimensions import predicate; predicate("10:P2")

— dimensions.py:507 (the check that decides 10:P2; family dom.D)
  507      # 10:P2 (p10054)
  508      # constancy (the prediction P2): the dimensionless ratios of the constants are fixed, zero drift; the present bound on the
  509      # fine-structure constant, (1/alpha) dalpha/dt = 1.8(2.5)e-19 /yr (Filzinger et al. 2023, imported), brackets the predicted 0:
  510      # 18 - 25 <= 0 <= 18 + 25 in units of 1e-20 /yr.
  511      chk("P2: the predicted drift 0 lies within the imported bound 1.8(2.5)e-19 /yr", 18 - 25 <= 0 <= 18 + 25)
  512  
  513      # 10:F3 (p10035), 10:X2 (p10049)
  514      # temperature and Unruh closures in (r, s, j) bookkeeping
  515      kB_dom = (-1, 1, 1)                    # Iq L^-1 T
  516      E_dom3 = (0, -1, 1)                    # Iq T^-1
  517      acc = (1, -2, 0)
  518      Theta = tuple(a - b for a, b in zip(E_dom3, kB_dom))
  519      chk("[Theta] = [E][k_B]^-1 = L T^-2, flag-free", Theta == acc)
  520      c_dom3 = (1, -1, 0)
  [PASS] dom.D    

In [24]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:X1 (p10048) [T] — the deciding family dom.B
# The arity of classical bookkeeping, derived: four horizons less one identity, $4-1=3$ free scales in
# bijection with $\{c,\hbar,G\}$, exactly three independent members of $\{c,\hbar,G,k_B\}$; the three
# mechanical primitives are the two chart generators and the flag's torsion-free surrogate, mass,
# $[m]=\Iq\unitL^{-2}\unitT$, $m_P=\pP/c$. Classical dimensional analysis postulates the count and the mass
# primitive.
from frc_10_dimensions import predicate; predicate("10:X1")

— dimensions.py:357 (the check that decides 10:X1; family dom.B)
  357      # 10:X1 (p10048)
  358      chk("DOF: 4 - 1 = 3 free scales", 4 - rank_int([rel]) == 3)
  359  
  360      # ------------------------------------------------------------------
  361      # flag positions of the quartet: (l, t, p, E) at (0, 0, 1, 1) mod 4;
  362      # every quartet relation must close on the flag component
  363      # ------------------------------------------------------------------
  364      FL = {"l": 0, "t": 0, "p": 1, "E": 1}
  365      def flag_of(expr):
  366          # expr: dict of generator -> exponent
  367          return sum(FL[g] * e for g, e in expr.items()) % 4
  368      chk("flag[hbar = l p] = 1", flag_of({"l": 1, "p": 1}) == 1)
  369      chk("flag[hbar = t E] = 1", flag_of({"t": 1, "E": 1}) == 1)
  370      chk("flag[c = l/t] = 0", flag_of({"l": 1, "t": -1}) == 0)
  [PASS] dom.A    (corroborating)  33 exact micro-checks
  [PASS] dom.B    (deciding)       30 exact micro-che

In [25]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:X2 (p10049) [T] — the deciding family dom.D
# The thermal primitive is redundant: $[\Theta]=\unitE[k_B]^{-1}=\unitL\unitT^{-2}$, the acceleration domain,
# flag-free where energy carries the flag; $|k_B|=\hbar/c$ is the quartet's mixed product; the Unruh
# combination $\hbar a/(c\,k_B)$ closes as a domain identity. Classical bookkeeping adjoins a thermal unit and
# reads $k_B$ as a conversion.
from frc_10_dimensions import predicate; predicate("10:X2")

— dimensions.py:513 (the check that decides 10:X2; family dom.D)
  513      # 10:F3 (p10035), 10:X2 (p10049)
  514      # temperature and Unruh closures in (r, s, j) bookkeeping
  515      kB_dom = (-1, 1, 1)                    # Iq L^-1 T
  516      E_dom3 = (0, -1, 1)                    # Iq T^-1
  517      acc = (1, -2, 0)
  518      Theta = tuple(a - b for a, b in zip(E_dom3, kB_dom))
  519      chk("[Theta] = [E][k_B]^-1 = L T^-2, flag-free", Theta == acc)
  520      c_dom3 = (1, -1, 0)
  521      hb_dom3 = (0, 0, 1)
  522      unruh = tuple(h + a - c - k for h, a, c, k in zip(hb_dom3, acc, c_dom3, kB_dom))
  523      chk("Unruh closure: [hbar a / (c k_B)] = [Theta]", unruh == acc)
  524  
  525      # crossing-degree recovery (Cor recovery-crossings)
  526      def embed(u, a, b):
  [PASS] dom.B    (corroborating)  30 exact micro-checks
  [PASS] dom.D    (deciding)       51 exact micro-checks
10:X2: VERIFIED — 2 family record(s)


In [26]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:X3 (p10050) [T] — the deciding family dom.A
# One action quantum for both conjugate pairs: the domain lattice has exactly one order-four subgroup, in the
# time factor (no flag of space), canonical and covariant, so $E=hf$ and $p=\hbar k$ share one $\hbar$,
# $\lP\pP=\tPl\EP$ --- derived from the shell's algebra with no Lorentz structure used.
from frc_10_dimensions import predicate; predicate("10:X3")

— dimensions.py:223 (the check that decides 10:X3; family dom.A)
  223      # 10:X3 (p10050)
  224      chk("unique order-four subgroup: exactly two generators", sorted(o4) == [(0, KAPPA), (0, 3 * KAPPA)])
  225      chk("no flag of space: all order-four elements have r = 0", all(r == 0 for r, _ in o4))
  226      chk("flag horizon-inaccessible on window H<kappa",
  227            all((0, KAPPA) != (r % P, s % (P - 1))
  228                for r in range(-1, 2) for s in range(-1, 2)))
  229      EPSILONS = [e for e in range(1, P - 1) if __import__("math").gcd(e, P - 1) == 1]
  230      chk("flag covariance: eps*kappa in {kappa, 3kappa} for all units eps",
  231            all((e * KAPPA) % (P - 1) in (KAPPA, 3 * KAPPA) for e in EPSILONS))
  232      chk("realization is a homomorphism",
  233            all(realize(dmul(a, b)) == ((realize(a)[0] + realize(b)[0]) % P,
  234                                        (realize(a)[1] + realize(b)[1]) % (P - 1))
  235                for a in [L,

In [27]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
# 10:X4 (p10051) [T] — the deciding family dom.G
# Charge is an integer degree of the two generators on the shells of odd capacity, not a further primitive:
# with the Coulomb constant neutral (F5), $[q]^{2}=\Iq\unitL\unitT^{-1}$, $q_P^{2}=\hbar c$; the lift has no
# half root (the Gaussian exponents $M^{1/2}L^{3/2}T^{-1}$ are half-integers there), so torsion-free
# bookkeeping adjoins an electrical primitive as it adjoins mass; the fine-structure ratio stays a neutral
# count.
from frc_10_dimensions import predicate; predicate("10:X4")

— dimensions.py:709 (the check that decides 10:X4; family dom.G)
  709      # 10:F5 (p10047), 10:X4 (p10051)
  710      chk("charge: [q]^2 = [E][L] = I_q [L][T]^-1 = (1, 2) on F_13, its roots exactly (7,1) and (7,7), differing by the half-period (0, 6)",
  711            ((0 + 1) % 13, (3 - 1 + 0) % 12) == (1, 2) and charge_roots(3) == [(7, 1), (7, 7)] and ((7 - 7) % 13, (7 - 1) % 12) == (0, 6))
  712      chk("charge: the roots exist iff kappa is odd, then exactly (2 kappa+1, (kappa-1)/2) and its half-period partner (every prime p = 4 kappa+1, kappa <= 30)",
  713            all((charge_roots(k) == sorted([(2 * k + 1, (k - 1) // 2), (2 * k + 1, (k - 1) // 2 + 2 * k)])) if k % 2 else charge_roots(k) == []
  714                for k in range(1, 31) if all((4 * k + 1) % d for d in range(2, 4 * k + 1))))
  715      chk("charge: the space exponent 2^-1 = 2 kappa+1 = -2 kappa is the meridian half-turn, |r| = 2 kappa, on p = 13, 29, 173, 229",
  716            all((2 * (2 * k + 1)) % (4 * k 

## Summary

The cells above are the paper's python-witnessed predicates; the whole script, family by family, is `python3 dimensions.py`
(13 family checks, 226 exact micro-checks, `results.json`), `python3 -m frc_10_dimensions` once installed.

In [28]:
!pip install -q frc-10-dimensions --find-links https://finitering.space/pkg/
from frc_10_dimensions import verify_all
assert verify_all(), "a family check failed"
print("all family checks pass")


SUMMARY: 13/13 checks passed (226 exact micro-checks)
all family checks pass
